pkg/api/login_oauth.go

```diff
@@ -243,6 +249,22 @@ func (hs *HTTPServer) OAuthLogin(ctx *m.ReqContext) {
 	ctx.Redirect(setting.AppSubUrl + "/")
 }
 
+func isRoleAssignable(currentRole m.RoleType, incomingRole m.RoleType) bool {
+	// role hierarchy
+	roleHierarchy := map[m.RoleType]int{
+		m.ROLE_VIEWER: 0,
+		m.ROLE_EDITOR: 1,
+		m.ROLE_ADMIN:  2,
+	}
+
+	// If the incoming role is less than ( less privilege ) than the currently assigned role ( more privilege ), skip this mapping.
+	if currentRole != "" && roleHierarchy[m.RoleType(incomingRole)] < roleHierarchy[currentRole] {
```

Comment:
The RoleType has an `Includes` method that can be used to check the hierarchy. Basically if the new role that is read is already included in the role that is assigned already then leave the original, if not assign the new one.

Here's the signature of the Includes method: `func (r RoleType) Includes(other RoleType) bool`

Reply:
If the Includes method is used instead of this custom function `isRoleAssignable` then the testing should move back to the Includes implementation on `pkg/models/org_user.go`, that is actually not tested atm :(

Reply:
I think this is useful to keep in mind, and as I refactor these additions for an upstream PR we maybe can use that or get suggestions from the grafana maintainers.  LDAP and some of the other auth connectors aren't using it.